crates/block-producer/src/psc.rs

```diff
@@ -0,0 +1,908 @@
+                            last_valid,
+                            last_submitted,
+                            last_confirmed
+                        );
+                        if let Some(ref sync_server) = self.context.block_sync_server_state {
+                            let mut sync_server = sync_server.lock().unwrap();
+                            sync_server.publish_revert(
+                                Revert::new_builder().number_hash(last_submitted_nh).build(),
+                            );
+                            for b in last_confirmed + 1..=last_valid {
+                                publish_local_block(&mut sync_server, &snap, b)?;
+                            }
+                        }
+
+                        {
+                            let mut local_cells_manager =
+                                self.context.local_cells_manager.lock().await;
+                            local_cells_manager.reset();
+                            for b in last_confirmed + 1..=last_valid {
+                                let deposits =
+                                    snap.get_block_deposit_info_vec(b).expect("deposit info");
+                                let deposits = deposits.into_iter().map(|d| d.cell());
+                                for c in deposits {
+                                    local_cells_manager.lock_cell(c.out_point());
+                                }
+                            }
+                            let new_tip = snap.get_last_valid_tip_block_hash()?;
+                            let mut mem_pool = self.context.mem_pool.lock().await;
+                            mem_pool
+                                .notify_new_tip(new_tip, &local_cells_manager)
+                                .await?;
+                        }
+                        self.local_count = last_valid - last_submitted;
+                        self.submitted_count = last_submitted - last_confirmed;
+                    } else {
+                        bail!(e);
+                    }
+                }
+            }
+        }
+    }
+}
+
+async fn run(mut state: &mut ProduceSubmitConfirm) -> Result<()> {
+    let mut submitting = false;
+    let mut submit_handle = spawn_abort_on_drop(async { anyhow::Ok(NumberHash::default()) });
+    let mut confirming = false;
+    let mut confirm_handle = spawn_abort_on_drop(async { anyhow::Ok(NumberHash::default()) });
+    let config = &state.context.psc_config;
+    let mut interval = tokio::time::interval(Duration::from_secs(config.block_interval_secs));
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+    let mut revert_local_signal = signal(SignalKind::user_defined1())?;
+    let mut revert_submitted_signal = signal(SignalKind::user_defined2())?;
+
+    loop {
+        if !submitting && state.local_count > 0 && state.submitted_count < config.submitted_limit {
```

Comment:
Why not confirm first then submit?

Reply:
The order should not matter here because we are just spawning tasks and won't change state yet.